akka23-common/src/main/scala/com/typesafe/conductr/akka/ConnectionHandler.scala

```diff
@@ -70,39 +76,89 @@ class ConnectionHandler extends AbstractConnectionHandler {
-
+  override def withConnectedRequest[T](payload: Option[HttpPayload])(handler: (Int, Map[String, Option[String]]) => Option[T])(implicit cc: CC): Future[Option[T]] = {
     payload.fold[Future[Option[T]]](Future.successful(None)) { p =>
-      val connection = cc.httpExt.outgoingConnection(p.getUrl.getHost, p.getUrl.getPort)
-      val url = p.getUrl
-      val urlStr = url.toString
-
-      val requestMethod = p.getRequestMethod match {
-        case "GET"     => Get(urlStr)
-        case "POST"    => Post(urlStr)
-        case "PUT"     => Put(urlStr)
-        case "PATCH"   => Patch(urlStr)
-        case "DELETE"  => Delete(urlStr)
-        case "OPTIONS" => Options(urlStr)
-        case "HEAD"    => Head(urlStr)
-      }
+      val connection = createConnection(p)
+      val request = createRequest(p)
 
-      val request = requestMethod
-        .addHeader(`User-Agent`(UserAgent))
-        .addHeader(Host(url.getHost))
-
-      val requestSource = Source.single(request)
+      Source.fromFuture(request)
         .via(connection)
         .map { response =>
-          thunk(
+          handler(
             response.status.intValue(),
             response.headers.foldLeft(Map.empty[String, Option[String]]) {
               case (m, header) => m.updated(header.name(), Some(header.value()))
             })
         }
-      requestSource.runWith(Sink.head)(cc.actorMaterializer)
+        .runWith(Sink.head)(cc.actorMaterializer)
+    }
+  }
+
+  def withConnectedRequest[T](payload: HttpPayload, body: Option[Future[RequestEntity]] = None)(handler: (Int, Map[String, Option[String]], ResponseEntity) => Future[T])(implicit cc: CC): Future[T] = {
+    val connection = createConnection(payload)
+    val request = createRequest(payload, body)
+
+    Source.fromFuture(request)
+      .via(connection)
+      .mapAsync(1) { response =>
+        handler(
+          response.status.intValue(),
+          response.headers.foldLeft(Map.empty[String, Option[String]]) {
+            case (m, header) => m.updated(header.name(), Some(header.value()))
+          },
+          response.entity)
+      }
+      .runWith(Sink.head)(cc.actorMaterializer)
+  }
+
+  // FIXME: Sink.ignore is not handling the closed connection event
+  //        Should we instead return the closed connection event so that the user can handle it.
+  //        If so should we add an additional function to the method signature or return the closed event as the return type, e.g. as
+  //        - Try[Unit]
+  //        - Boolean (true = expected connection closing, false = connection closed due to failure)
+  //        - Custom object
+  //        Note, that returning the connection closed event as a return type could be irritating.
+  //        The user might think that the return type contains the actual events (e.g. SSE events).
+  def withConnectedStream[T](payload: HttpPayload)(unmarshalling: HttpResponse => Future[Source[T, Unit]])(implicit cc: CC): Future[Source[T, Unit]] = {
```

Comment:
Are these comments still relevant? It doesn't feel like we're consuming all of the stream here - which is what should do for any request being made (it is essential). We're using `response.entity.dataBytes.runWith(Sink.ignore)` in contrail, so if there's a problem then I'd like to understand it.

Actually, perhaps that is the issue you've uncovered i.e. not consuming the response entity's data entirely.